### DISCUSSION TITLE
Why does the memory increase so quickly during the calculation when the LTS Clustering value is large?

### DISCUSSION THREAD

**Discussion by buzhiqiuyuebai** [SeisSol/SeisSol#948] [Q&A] [answered]
The example I use is [here](https://github.com/buzhiqiuyuebai/MemoryIncreaseProblem).
Why is it that just over two thousand grids, as the simulation progresses, take up more than 20 G of memory, does it cause by 21 LTS clusterings?
![image](https://github.com/SeisSol/SeisSol/assets/114474894/3aa14505-5451-4461-8628-9b1b5b7c9baf)

**Comment by krenzland** [comment]
Can you please clarify what your issue is? How do you compute the memory usage?

**Comment by davschneller** [comment]
Ok, I may have reproduced the behavior, at least when extrapolating: running the given setup showed a steady increase in memory consumption (the testing methodology was in this case to check the resident memory size (RES) in htop for the SeisSol processes and watch it grow—or not). That was the case on the current `master` branch as of writing this message.

The reason for that seem to be the actor statistics, since the given setup causes the actors to switch state rather often, and thus accumulating data there quickly. In particular, it seems to be https://github.com/SeisSol/SeisSol/blob/6366ff9d2d6f9426b785d828d591aa8e3cf785d9/src/Monitoring/LoopStatistics.h#L82 and https://github.com/SeisSol/SeisSol/blob/6366ff9d2d6f9426b785d828d591aa8e3cf785d9/src/Monitoring/ActorStateStatistics.h#L20 ; once these two are commented out, there is barely any increase visible anymore—at least short-term (whereas with even one of these lines in, the count was steadily rising every few seconds).

(I had also run valgrind with `--leak-check=full` for a bit, and at least as far as I got, I couldn't see any memory leaks within the loops)

**Reply by buzhiqiuyuebai** [reply]
Thank you for your reply, I will have a try.

**Reply by Thomas-Ulrich** [reply]
I suppose these lines have to be there and cannot just be removed right? So David found the cause but not a solution to your problem (I guess this answers the other issue you opened during the week end).

**Reply by davschneller** [reply, marked as answer]
At least temporarily, they can be commented out—they only concern "telemetry" about the cluster behavior (what is printed at the end, how long we took in one kernel on average, the standard deviation and so on and so forth). The `actorStateStatistics` are to my knowledge only used in the `TimeCluster`, and likewise the `loopStatistics` (which are also declared in the `TimeManager`)—and both don't influence the actual behavior of these classes, to my knowledge at least. Especially, there shouldn't be much of a connection to Ghost clusters.

Thus, I'm not sure if this issue is related to #949

**Reply by krenzland** [reply]
You can remove them or better add them behind a compilation flag. In hindsight, it was a bad idea to add them to the default build.

**Reply by krenzland** [reply]
For most normal setups they shouldn't add much overhead though.

**Reply by davschneller** [reply]
I mean—having runtime infos on the kernels is a nice thing IMO.

What we could do is maybe accumulate already during the simulation? I.e. We may at least cut the memory usage in half by removing the `samples` vector in the Actor statistics, and add directly to the loop statistics class instead.

Secondly, the loop statistics then should only save explicit samples if we want to output them—otherwise, accumulate them locally immediately, to only output the regression analysis at the end. I.e. we run https://github.com/SeisSol/SeisSol/blob/6366ff9d2d6f9426b785d828d591aa8e3cf785d9/src/Monitoring/LoopStatistics.cpp#L58-L106 already locally after each sample—maybe with or without output.

**Reply by krenzland** [reply]
Yes, that'd be ideal. To be honest, I wasn't able to get that much useful information out of these cluster statistics anyway.

Averaging sounds good. For most use cases, we can average between sync points. That's what I ended up doing. This is enough to reliably compute the load imbalance between ranks and to compute the relative cost of the DR computations.

**Reply by krenzland** [reply]
However, in this case, the problem is pretty certainly caused by the mesh which seems to have quite poor quality.

**Reply by davschneller** [reply]
PR #952 should solve the issue, unless we enable the loop statistics output to Netcdf. But even then—the `ActorStateStatistics` now doesn't keep an extra sample vector anymore.

**Reply by buzhiqiuyuebai** [reply]
I tried what you did and it worked, does this change make a difference to the simulation results?

**Reply by davschneller** [reply]
It doesn't alter them at all; the disabled functionality is merely meant as a runtime analysis.

For example, consider the following line (it's from a testing run of our CI):

https://gitlab.lrz.de/ci_seissol/SeisSol/-/jobs/6650156#L272

The marked line (272) and the following show a small regression analysis, and it shows about 2736 samples (each sample is triggered by the line in `LoopStatistics.h`). In your simulation, you should see something like that near the end, but showing 0 samples there.

And that's basically all that changed in this case—merely some terminal output at the end of the simulation. You can safely keep these two lines commented out for now, if they should cause trouble again.

**Comment by buzhiqiuyuebai** [comment]
> Ok, I may have reproduced the behavior, at least when extrapolating: running the given setup showed a steady increase in memory consumption (the testing methodology was in this case to check the resident memory size (RES) in htop for the SeisSol processes and watch it grow—or not). That was the case on the current `master` branch as of writing this message.
> 
> The reason for that seem to be the actor statistics, since the given setup causes the actors to switch state rather often, and thus accumulating data there quickly. In particular, it seems to be
> 
> https://github.com/SeisSol/SeisSol/blob/6366ff9d2d6f9426b785d828d591aa8e3cf785d9/src/Monitoring/LoopStatistics.h#L82
> 
> and
> https://github.com/SeisSol/SeisSol/blob/6366ff9d2d6f9426b785d828d591aa8e3cf785d9/src/Monitoring/ActorStateStatistics.h#L20
> 
> ; once these two are commented out, there is barely any increase visible anymore—at least short-term (whereas with even one of these lines in, the count was steadily rising every few seconds).
> (I had also run valgrind with `--leak-check=full` for a bit, and at least as far as I got, I couldn't see any memory leaks within the loops)
